ecosystem/typescript/sdk/examples/typescript-esm/offer_capabilities.ts

```diff
@@ -0,0 +1,153 @@
+    console.log("\nChecking Alice's account resources to verify the rotation capability offer is for Bob...");
+    const { data } = await provider.getAccountResource(alice.address(), "0x1::account::Account");
+    const offerFor = (data as any).rotation_capability_offer.for.vec[0];
+
+    console.log({ rotation_capability_offer: { for: offerFor } });
+    assert(offerFor.toString() == bob.address().toString(), "Bob's address should be in the rotation capability offer");
+    console.log("...success!\n");
+  }
+
+  // Offer Alice's signer capability to Bob
+  {
+    // Construct the SignerCapabilityOfferProofChallengeV2 struct
+    const signerCapProof: CapabilityOfferProofChallengeV2 = {
+      accountAddress: moduleAddress,
+      moduleName: "account",
+      structName: "SignerCapabilityOfferProofChallengeV2",
+      sequenceNumber: Number((await provider.getAccount(alice.address())).sequence_number), // Get latest sequence number
+      sourceAddress: aliceAccountAddress,
+      recipientAddress: bobAccountAddress,
+      // Note no chainId, the signer capability offer doesn't require it. We leave it undefined
+    };
+
+    console.log(`\n---------------  SignerCapabilityOfferProofChallengeV2 ---------------\n`);
+
+    // Sign the BCS-serialized struct, submit the transaction, and wait for the result.
+    const res = await signStructAndSubmitTransaction(provider, alice, "offer_signer_capability", signerCapProof);
+
+    // Print the relevant transaction submission info
+    const { hash, version, success, payload } = res;
+    console.log("Submitted transaction results:");
+    console.log({ hash, version, success, payload });
+
+    // Query Alice's Account resource on-chain to verify that she has offered the signer capability to Bob
+    console.log("\nChecking Alice's account resources to verify the signer capability offer is for Bob...");
+    const { data } = await provider.getAccountResource(alice.address(), "0x1::account::Account");
+    const offerFor = (data as any).signer_capability_offer.for.vec[0];
+
+    console.log({ signer_capability_offer: { for: offerFor } });
+    assert(offerFor.toString() == bob.address().toString(), "Bob's address should be in the signer capability offer\n");
+    console.log("...success!\n");
+  }
+})();
+
+const signStructAndSubmitTransaction = async (
+  provider: Provider,
+  signer: AptosAccount,
+  funcName: string,
+  struct: CapabilityOfferProofChallengeV2,
+): Promise<any> => {
+  // The proof bytes are just the individual BCS serialized
+  // data concatenated into a single byte array.
+  // Note that the proof bytes must be constructed in this specific order: the order of the struct data on-chain.
+  const proofBytes = new Uint8Array([
+    ...BCS.bcsToBytes(struct.accountAddress),
+    ...BCS.bcsSerializeStr(struct.moduleName),
+    ...BCS.bcsSerializeStr(struct.structName),
+    ...(struct.chainId ? BCS.bcsSerializeU8(struct.chainId) : []),
```

Comment:
this is tricky, because the ideal way to write this code is that this is a serialize function on the struct

Reply:
Yeah I realized there's actually already a struct for the rotation offer in the sdk, and then after seeing that + working with Kent on it I realized it's pretty janky to do it this way rather than the struct being a class that implements serialize

I'll change it to that before submitting